nixos/tests/invidious.nix

```diff
@@ -8,6 +8,7 @@ import ./make-test-python.nix ({ pkgs, ... }: {
   nodes.machine = { config, lib, pkgs, ... }: {
     services.invidious = {
       enable = true;
+      settings.hmac_key = "required-key";
```

Comment:
How does the NixOS module behave when this isn't specified? Does it error at eval-time, build-time or run-time?

Reply:
I would say that nix will not complain (at any stage) if `hmac_key` is not specified.
Maybe we can add an assertion to check if it is indeed provided by the user or not.

Reply:
Invidious will complain at run time if it's missing. There isn't really an easy way to check this though as `hmac_key` is supposed to be secret, so should be configured as raw json in `services.invidious.extraSettingsFile`.

Reply:
Ok, then the only thing we could do is warn the user if neither `services.invidious.settings.hmac_key` or `services.invidious.extraSettingsFile` is not set.
But it might not even be our responsibility to make sure that someone's invidious config is correct or not.

Reply:
It's a really nice property for the service to just work by setting `enable = true`. Can the secret be generated at runtime? Can it change every time it starts or should it be persistent? If the latter it can be stored in `/var/lib/invidious` somewhere. And is there a reason a user would want to not use an auto-generated default?

Reply:
yes I agree on the "work OOTB" argument.

[In the documentation](https://docs.invidious.io/installation/#post-install-configuration) it is stated that it can be generated randomly using the command `pwgen 20 1`.
[In the docker file](https://github.com/iv-org/documentation/pull/415/files), they expect the user to provide it.
Maybe it is fine if we generate it at build-time.

Reply:
Definitely shouldn't be done at build-time, otherwise the secret would end up in the Nix store. Should be run-time instead. If we had something like https://github.com/NixOS/nixpkgs/pull/179471 this would be easier to do, still possible without it though

Reply:
Any thoughts on how to add an hmac default without also breaking existing nixos module configs? Mine looks like this today:

```nix
invidious = {
  enable = true;
  package = unstable.invidious; # unstable-2023-07-05, already requires hmac_key
  settings = {
    # many things that are not hmac_key
  };
  extraSettingsFile = "/etc/invidious/config.yml"; # contains hmac_key
};
```

Obviously I can switch to doing whatever myself, but without checking extraSettingsFile for hmac_key's presence I'm not sure how you avoid stepping on toes for current users of unstable. Also not sure how big a concern it should be relative to friendliness to new people adding for the first time.

Reply:
The HMAC is used for app-level auth and CSRF tokens per https://github.com/iv-org/invidious/blob/master/config/config.example.yml#L437. I believe if it changes then you get some client credentials invalidated and errors thrown for existing sessions, users having to log out/clear cookies then log back in, etc. So you do want it to persist if possible, which is why they changed it to a mandatory config.

Reply:
Oh I wasn't aware of `extraSettingsFile`, I didn't have that in my [original PR](https://github.com/NixOS/nixpkgs/pull/67664), and I don't think it should've been introduced, especially in light of NixOS/rfcs#42.

Instead it would've been best to have a `hmacKeyFile` option of type `nullOr path`, defaulting to `null`, implemented something like this:
```nix
{
  services.invidious.settings.hmac_key = lib.mkIf (cfg.hmacKeyFile != null) (lib.mkDefault "NIXOS_HMAC_KEY_AUTO_REPLACED");

  systemd.services.invidious.preStart =
    if cfg.hmacKeyFile != null then ''
      hmacKey=$(<${lib.escapeShellArg cfg.hmacKeyFile})
      sed "s/NIXOS_HMAC_KEY_AUTO_REPLACED/$hmacKey/" ${configFile} > config/config.yml
    '' else ''
      cp ${configFile} config/config.yml
    '';
}
```

Then for this update making requiring, `hmac_key` required, it could've easily been changed to this:
```nix
{
  services.invidious.settings.hmac_key = lib.mkDefault "NIXOS_HMAC_KEY_AUTO_REPLACED";

  systemd.services.invidious.preStart = ''
    ${if cfg.hmacKeyFile != null then ''
      hmacKeyFile=${lib.escapeShellArg cfg.hmacKeyFile}
    '' else ''
      hmacKeyFile=/var/lib/invidious/hmac_key
      if [[ ! -f "$hmacKeyFile" ]]; then
        # TODO: Ensure restrictive permissions
        pwgen 20 1 > "$hmacKeyFile"
      fi
    ''}
    hmacKey=$(<${lib.escapeShellArg cfg.hmacKeyFile})
    sed "s/NIXOS_HMAC_KEY_AUTO_REPLACED/$hmacKey/" ${configFile} > config/config.yml
  '';
}
```

Reply:
I would still be in favor of deprecating the `extraSettingsFile` option, mainly because it hurts the ability for other modules to inspect values via `settings`.

However, it turns out that `extraSettingsFile` is implemented nicely by merging it with the generated `settings` file using a `jq` filter: https://github.com/NixOS/nixpkgs/blob/c25960f7e89535622359790c03bcd9abfed7c298/nixos/modules/services/web-apps/invidious.nix#L19-L26

Which means we can easily extend the filter to merge it with another file. In particular such that `extraSettingsFile` takes precedence over `settings`, which takes precedence over an autogenerated config file with an `hmac_key`.
It could even be checked using `jq` whether `hmac_key` is already set anywhere before even generating the `hmac_key`!

Reply:
I've got a mostly working version for this now. I'll test some more things and then push it, most likely tomorrow.